add server to filtered cipher suites log

ramaraochavali opened the pull request:
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure


Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

ramaraochavali commented:
@howardjohn would this work?

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 5803b16090e71bae5702c33f1005af400aae3094 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/lint_istio/1417449869764726784) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test lint_istio

hzxuzhonghu commented on pilot/pkg/model/gateway.go:
```diff
@@ -167,6 +168,13 @@ func MergeGateways(gateways []gatewayWithInstances) *MergedGateway {
+					if len(s.Tls.CipherSuites) > 0 {
+						serverName := s.Name
+						if len(serverName) == 0 {
+							serverName = s.Port.Name
```
This is hacky

ramaraochavali replied:
So we either identify server with unique port name or server name. Why do you think it is hacky?

hzxuzhonghu reviewed (commented):
lgtm

hzxuzhonghu commented:
Just noticed the issue. For the previous gateway with invalid cipher suites, it is invalid with nack. I would rather prevent it in validation. We can assume no such gateways exist in mesh.

ramaraochavali commented:
https://github.com/istio/istio/pull/34109 - We discussed about that here and decided to not prevent such gateways but add warning because a gateway may have many servers and only a single cipher suite may be invalid in server.

howardjohn commented on pilot/pkg/model/gateway.go:
```diff
@@ -167,6 +168,13 @@ func MergeGateways(gateways []gatewayWithInstances) *MergedGateway {
+						if len(serverIdentifier) == 0 {
+							serverIdentifier = s.Port.Name
+						}
+						s.Tls.CipherSuites = filteredCipherSuites(s.Tls.CipherSuites, serverIdentifier, gatewayName)
```
is it safe to mutate here? Will this impact the underlying client-go cache?

hzxuzhonghu replied:
I didnot realize it is mutating the original object

ramaraochavali replied:
So is it OK if filteredCipherSuite returns a different set, we clone the server and set?

hzxuzhonghu replied:
What is this pr fixing?

ramaraochavali replied:
#34109 added validation and logic to skip invalid cipher suites. But the logic to skip is executed on every listener creation. I am trying to see if we can do it once with this PR.

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 13cdbd98c0a40ed027f720138d954e9fce46d260 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/integ-pilot-k8s-tests_istio/1417746129801449472) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 13cdbd98c0a40ed027f720138d954e9fce46d260 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/integ-pilot-k8s-tests_istio/1417746129801449472) | `/test integ-pilot-k8s-tests_istio`
integ-cni-k8s-tests_istio | 13cdbd98c0a40ed027f720138d954e9fce46d260 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/integ-cni-k8s-tests_istio/1417746129839198208) | `/test integ-cni-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-pilot-k8s-tests_istio

ramaraochavali commented:
/test integ-cni-k8s-tests_istio

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 13cdbd98c0a40ed027f720138d954e9fce46d260 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/integ-pilot-k8s-tests_istio/1417746129801449472) | `/test integ-pilot-k8s-tests_istio`
integ-cni-k8s-tests_istio | 13cdbd98c0a40ed027f720138d954e9fce46d260 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/integ-cni-k8s-tests_istio/1417746129839198208) | `/test integ-cni-k8s-tests_istio`
integ-pilot-multicluster-tests_istio | 13cdbd98c0a40ed027f720138d954e9fce46d260 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/integ-pilot-multicluster-tests_istio/1417746131416256512) | `/test integ-pilot-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
I tried to clone when ciphersuites are invalid - but selective cloning does not work as the same server is referenced in multiple places like serversByRouteName. I felt cloning every server is not efficient - rather we pay the cost of comparison at listener creation. So I reverted that change and just added server to log to indicate which server is causing issue.

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | 57b7e5392798b8a7c4d6850102bf0e896061c8b1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/integ-pilot-k8s-tests_istio/1417763401311981568) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-pilot-k8s-tests_istio

howardjohn commented on pilot/pkg/networking/core/v1alpha3/gateway.go:
```diff
@@ -992,3 +979,17 @@ func buildGatewayVirtualHostDomains(hostname string, port int) []string {
+		if security.IsValidCipherSuite(s) {
+			ret = append(ret, s)
+		} else {
+			log.Debugf("ignoring unsupported cipherSuite: %q for server %s", s, server.String())
```
nit: server.String is fairly expensive. Maybe we can wrap in `if log.DebugEnabled` so we do not pay for it when debug log is off

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 57955973e00bffe14882ce57b8e45ca8f03d75b1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/lint_istio/1418072858604605440) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 57955973e00bffe14882ce57b8e45ca8f03d75b1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/lint_istio/1418072858604605440) | `/test lint_istio`
integ-telemetry-mc-k8s-tests_istio | 57955973e00bffe14882ce57b8e45ca8f03d75b1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/34170/integ-telemetry-mc-k8s-tests_istio/1418072853802127360) | `/test integ-telemetry-mc-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
